controllers/keystoneapi_controller.go

```diff
@@ -109,6 +109,26 @@ func (r *KeystoneAPIReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return ctrl.Result{}, err
 	}
 
+	helper, err := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		r.Log,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if instance.DeletionTimestamp.IsZero() && !controllerutil.ContainsFinalizer(instance, helper.GetFinalizer()) {
+		// If the service object doesn't have our finalizer, add it.
+		controllerutil.AddFinalizer(instance, helper.GetFinalizer())
+		// Register the finalizer immediately to avoid orphaning resources on delete
+		err := r.Update(ctx, instance)
+
+		return ctrl.Result{}, err
```

Comment:
instead of use `controllerutil.ContainsFinalizer` could do this in one call and check the return of `controllerutil.AddFinalizer`:
~~~
It returns an indication of whether it updated the object's list of finalizers.
~~~

```
// Register the finalizer immediately to avoid orphaning resources on delete
if controllerutil.AddFinalizer(instance, helper.GetFinalizer()) {
  err := r.Update(ctx, instance)
  return ctrl.Result{}, err
}
```

Reply:
Ah, I had never read the function description.  That sounds good and I will change that.